File: backend/app/schemas/notifications.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class NotificationResponse(BaseModel):
    id: int
    module_key: str
    category: str
    severity: str
    title: str
    body: str
    link_url: str
    read_at: datetime | None
    created_at: datetime
    expires_at: datetime | None


class NotificationListResponse(BaseModel):
    items: list[NotificationResponse]
    unread_count: int


class NotificationSettingUpdate(BaseModel):
    in_app_enabled: bool | None = None
    push_enabled: bool | None = None
    daily_digest_enabled: bool | None = None
    daily_digest_time: str | None = Field(default=None, pattern=r"^\d{2}:\d{2}$")
    due_soon_enabled: bool | None = None
    due_soon_hours: int | None = Field(default=None, ge=1, le=168)
    approval_notifications_enabled: bool | None = None
    quiet_hours_start: str | None = Field(default=None, pattern=r"^\d{2}:\d{2}$")
    quiet_hours_end: str | None = Field(default=None, pattern=r"^\d{2}:\d{2}$")


class NotificationSettingResponse(BaseModel):
    module_key: str
    settings: dict[str, Any]


class PushConfigResponse(BaseModel):
    vapid_public_key: str


class PushSubscriptionKeys(BaseModel):
    p256dh: str = Field(min_length=1, max_length=2048)
    auth: str = Field(min_length=1, max_length=512)


class PushSubscriptionCreate(BaseModel):
    endpoint: str = Field(min_length=1, max_length=2000)
    keys: PushSubscriptionKeys
    device_label: str = Field(default="", max_length=255)


class PushSubscriptionResponse(BaseModel):
    id: int
    endpoint: str
    device_label: str
    enabled: bool
    created_at: datetime
    last_seen_at: datetime

File: packages/family-api/src/models/notifications.ts
export interface NotificationItem {
  id: number;
  module_key: string;
  category: string;
  severity: string;
  title: string;
  body: string;
  link_url: string;
  read_at: string | null;
  created_at: string;
  expires_at: string | null;
}

export interface NotificationListResponse {
  items: NotificationItem[];
  unread_count: number;
}

export interface NotificationSettings {
  in_app_enabled: boolean;
  push_enabled: boolean;
  daily_digest_enabled: boolean;
  daily_digest_time: string;
  due_soon_enabled: boolean;
  due_soon_hours: number;
  approval_notifications_enabled: boolean;
  quiet_hours_start: string;
  quiet_hours_end: string;
}

export type NotificationSettingsByModule = Record<string, NotificationSettings>;

export type NotificationSettingUpdate = Partial<NotificationSettings>;

export interface NotificationSettingResponse {
  module_key: string;
  settings: NotificationSettings;
}

export interface PushConfigResponse {
  vapid_public_key: string;
}

export interface PushSubscriptionCreate {
  endpoint: string;
  keys: {
    p256dh: string;
    auth: string;
  };
  device_label?: string;
}

export interface PushSubscriptionResponse {
  id: number;
  endpoint: string;
  device_label: string;
  enabled: boolean;
  created_at: string;
  last_seen_at: string;
}
